pkg/daemon/certificate_writer.go

```diff
@@ -0,0 +1,86 @@
+package daemon
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (dn *Daemon) handleControllerConfigEvent(obj interface{}) {
+	controllerConfig := obj.(*mcfgv1.ControllerConfig)
+	glog.V(4).Infof("Updating ControllerConfig %s", controllerConfig.Name)
+	dn.enqueueControllerConfig(controllerConfig)
+}
+
+func (dn *Daemon) enqueueControllerConfig(controllerConfig *mcfgv1.ControllerConfig) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(controllerConfig)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %w", controllerConfig, err))
+		return
+	}
+	dn.ccQueue.AddRateLimited(key)
+}
+
+func (dn *Daemon) controllerConfigWorker() {
+	for dn.processNextControllerConfigWorkItem() {
+	}
+}
+
+func (dn *Daemon) processNextControllerConfigWorkItem() bool {
+	key, quit := dn.ccQueue.Get()
+	if quit {
+		return false
+	}
+	defer dn.ccQueue.Done(key)
+
+	err := dn.syncControllerConfigHandler(key.(string))
+	dn.handleControllerConfigErr(err, key)
+
+	return true
+}
+
+func (dn *Daemon) handleControllerConfigErr(err error, key interface{}) {
+	if err == nil {
+		dn.ccQueue.Forget(key)
+		return
+	}
+
+	if err := dn.updateErrorState(err); err != nil {
+		glog.Errorf("Could not update annotation: %v", err)
+	}
+	// This is at V(2) since the updateErrorState() call above ends up logging too
+	glog.V(2).Infof("Error syncing ControllerConfig %v (retries %d): %v", key, dn.ccQueue.NumRequeues(key), err)
+	dn.ccQueue.AddRateLimited(key)
+}
+
+func (dn *Daemon) syncControllerConfigHandler(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing ControllerConfig %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing ControllerConfig %q (%v)", key, time.Since(startTime))
+	}()
+
+	if key != ctrlcommon.ControllerConfigName {
+		// In theory there are no other ControllerConfigs other than the machine-config one
+		// but to future-proof just in case, we don't need to sync on other changes
+		return nil
+	}
+
+	controllerConfig, err := dn.ccLister.Get(ctrlcommon.ControllerConfigName)
+	if err != nil {
+		return fmt.Errorf("Could not get ControllerConfig: %v", err)
+	}
+
+	kubeAPIServerServingCABytes := controllerConfig.Spec.KubeAPIServerServingCAData
+
+	if err := writeFileAtomicallyWithDefaults(caBundleFilePath, kubeAPIServerServingCABytes); err != nil {
```

Comment:
This means we're rewriting the file even if it hasn't changed, but something else in the controllerconfig did?  That may be OK for now.  But I am sure if this approach was to be extended at all beyond this we'd want to enhance things.

I've been thinking about things like this in the context of https://github.com/containers/bootc/issues/22 and if we represented this certificate as a bootc configmap, the design I have in mind for the bootc configmap support would use ostree as a backend, and inherently hash its inputs.

Reply:
I can certainly make it more nuanced. This is more of a hacky way to get us past the issues highlighted in https://issues.redhat.com/browse/MCO-499, with a better way to do it in 4.14